pkg/geo/geomfn/node.go

```diff
@@ -0,0 +1,124 @@
+// Copyright 2020 The Cockroach Authors.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package geomfn
+
+import (
+	"fmt"
+
+	"github.com/cockroachdb/cockroach/pkg/geo"
+	"github.com/twpayne/go-geom"
+	"github.com/twpayne/go-geom/xy"
+)
+
+// extractUniqueEndpoints extracts the endpoints from geometry and removes duplicates
+func extractUniqueEndpoints(g geo.Geometry) (geo.Geometry, error) {
+	endpoints, err := extractEndpoints(g)
+	if err != nil {
+		return geo.Geometry{}, fmt.Errorf("error extracting endpoints: %v", err)
+	}
+
+	output, err := UnaryUnion(endpoints)
+	if err != nil {
+		return geo.Geometry{}, fmt.Errorf("error removing duplicates: %v", err)
+	}
+	return output, nil
+}
+
+// extractEndpoints extracts the endpoints from geometry provided and returns them as a multipoint geometry
+func extractEndpoints(g geo.Geometry) (geo.Geometry, error) {
+	mp := geom.NewMultiPoint(geom.XY)
+
+	gt, err := g.AsGeomT()
+	if err != nil {
+		return geo.Geometry{}, fmt.Errorf("error transforming geometry: %v", err)
+	}
+
+	switch gt := gt.(type) {
+	case *geom.LineString:
+		if err := collectEndpoints(gt, mp); err != nil {
+			return geo.Geometry{}, fmt.Errorf("error collecting endpoints: %v", err)
+		}
+	case *geom.MultiLineString:
+		for i := 0; i < gt.NumLineStrings(); i++ {
+			ls := gt.LineString(i)
+			if err := collectEndpoints(ls, mp); err != nil {
+				return geo.Geometry{}, fmt.Errorf("error collecting endpoints: %v", err)
+			}
+		}
+	default:
+		return geo.Geometry{}, fmt.Errorf("unsupported type: %T", gt)
+	}
+
+	result, err := geo.MakeGeometryFromGeomT(mp)
+	if err != nil {
+		return geo.Geometry{}, fmt.Errorf("error creating output geometry: %v", err)
+	}
+	return result, nil
+}
+
+// collectEndpoints takes endpoints of the line provided and adds them to the provided multipoint
```

Comment:
nit: period at end of sentence.

I think this function should return a list of `geom.Coord`, and the calling function should add it to the existing MultiPoint collection to make this function purer with no side effects.

Reply:
Makes sense, but should not it return a list of `geom.Point`s?

Reply:
either works in this case.